src/regression-lib/regression-tool.ts

```diff
@@ -0,0 +1,499 @@
+#!/usr/bin/env node
+
+import * as fs from 'fs';
+import chalk from 'chalk';
+import { ConstructAnalyzer } from '../library/analyzer';
+import { CONFIG } from '../library/config';
+import type { SignalWeights } from '../library/types';
+
+/**
+ * Input format: map of package names to target scores
+ */
+interface TargetScores {
+  [packageName: string]: number;
+}
+
+/**
+ * Output format: map of signal names to optimal weights
+ */
+type OptimalWeights = SignalWeights;
+
+/**
+ * Training data point for regression
+ */
+interface TrainingPoint {
+  packageName: string;
+  targetScore: number;
+  signalValues: Record<string, number>;
+}
+
+/**
+ * Experimental CLI tool for generating optimal signal weights using regression
+ */
+class RegressionCLI {
+  private analyzer: ConstructAnalyzer;
+  private allSignalNames: string[];
+
+  constructor() {
+    this.analyzer = new ConstructAnalyzer();
+    this.allSignalNames = this.extractAllSignalNames();
+  }
+
+  /**
+     * Extract all signal names from the current config
+     */
+  private extractAllSignalNames(): string[] {
+    const signalNames: string[] = [];
+    for (const pillar of CONFIG.pillars) {
+      for (const signal of pillar.signals) {
+        signalNames.push(signal.name);
+      }
+    }
+    return signalNames;
+  }
+
+  /**
+     * Main entry point for the CLI
+     */
+  async run(inputFile?: string): Promise<void> {
+    try {
+      console.log(chalk.bold.cyan('Experimental Signal Weight Regression Tool'));
+      console.log(chalk.cyan('============================================\n'));
+
+      // Load target scores
+      const targetScores = await this.loadTargetScores(inputFile);
+      console.log(chalk.green(`Loaded ${Object.keys(targetScores).length} packages with target scores`));
+
+      // Collect training data
+      console.log(chalk.blue('Analyzing packages to collect signal data...'));
+      const trainingData = await this.collectTrainingData(targetScores);
+
+      // Perform regression
+      console.log(chalk.blue('Performing regression to find optimal weights...'));
+      const optimalWeights = this.performRegression(trainingData);
+
+      // Output results
+      await this.outputResults(optimalWeights, trainingData);
+
+    } catch (error) {
+      console.error(chalk.red('Error:'), error instanceof Error ? error.message : String(error));
+      process.exit(1);
+    }
+  }
+
+  /**
+     * Load target scores from file or prompt user
+     */
+  private async loadTargetScores(inputFile?: string): Promise<TargetScores> {
```

Comment:
can you check why the _colors_ are messed up from this point onwards? `private` is white whereas above this its colored `blue`. I think its because typescript generally expects all `private` functions to be defined _above_ any public functions but honestly im not quite sure